fix: return proper error responses in streaming instead of silently closing streams [ARKQB-415]

av3n93rz opened the pull request:
## Checklist

- [ ] Follow the [contributor guide](./CONTRIBUTING.md)
- [ ] End-to-end tests pass
- [ ] Unit tests for essential parts of your code, e.g. APIs exposed via SDKs or endpoints
- [ ] Update `./docs`
- [ ] Recognize contributors with "@all-contributors please add <person> for <code, bug, docs, etc>"
- [ ] Linked issues #eg101

av3n93rz committed:
feat: sets HTTP 408 status on stream timeout response

Ensures proper indication of request timeout by sending HTTP 408 status when no stream chunks are received within the expected period.

Improves client-side error handling and clarity for timeout scenarios.

av3n93rz committed:
feat: improves error handling for streaming response failures

Enhances the proxy logic to extract and forward detailed error messages from failed streaming responses, ensuring clients receive more informative SSE error events. Helps with debugging and improves user feedback in case of upstream errors.

av3n93rz committed:
feat: improves error streaming and handling in chat workflows

Enables structured error chunks to be streamed alongside metadata, allowing clients to handle agent errors more gracefully. Updates backend and frontend to parse and relay error details in streaming endpoints for clearer feedback and better API reliability.

av3n93rz committed:
refactor: extract metadata building logic into separate function

dwmkerr commented on tools/ark-cli/src/lib/arkApiClient.ts:
```diff
@@ -158,13 +158,18 @@ export class ArkApiClient {
+      for await (const chunk of stream) {
+        yield chunk;
+      }
+    } catch (err) {
```
I *think* this will already show a proper error message to the user and kill the CLI so we can remove the try/catch and just a comment

dwmkerr commented on services/ark-api/ark-api/src/ark_api/api/v1/openai.py:
```diff
@@ -87,22 +87,50 @@ def process_request_metadata(
     return None
 
 
+# TODO: Maybe we should start streaming first, wait for the first chunk/response
```
can you double check this - either fix it or maybe raise a GH issue suggesting the improvement and then link to the GH issue rather than `TODO`

dwmkerr commented on services/ark-api/ark-api/src/ark_api/api/v1/openai.py:
```diff
@@ -87,22 +87,50 @@ def process_request_metadata(
     async with httpx.AsyncClient(timeout=timeout) as client:
         async with client.stream("GET", streaming_url) as response:
             if response.status_code != 200:
+                # Try to extract error message from response body
```
given we control the error structure, let's jsut read it with the expected structure and then fail otherwise, rather than trying fallbacks etc

dwmkerr commented on services/ark-api/ark-api/src/ark_api/api/v1/openai.py:
```diff
@@ -87,22 +87,50 @@ def process_request_metadata(
+                    pass
+
+                # Forward the error response as an SSE error event
+                error_data = {
```
nice, maybe we extract to a type perhaps same w/ incoming error? and then we have a few streaming types [perhaps useful to put them in their own file - cause we'll likely use elsewhere later]

dwmkerr commented on services/ark-cluster-memory/ark-cluster-memory/src/routes/stream.ts:
```diff
@@ -22,6 +25,30 @@ const writeSSEEvent = (res: Response, data: StreamResponse): boolean => {
+  let status = 500
+  let message = ''
+
+  if (typeof error === "string") {
```
again, given we know the structure can we use it and then fail if not? rather than trying fallbacks

dwmkerr reviewed (changes requested):
looks good, minor comments

lucaromagnoli committed:
fix: stream errors properly in SSE format instead of closing stream

- Fix StreamingError to use OpenAI-compatible error format with error.message, error.type, error.code
- Stream error chunks when query execution fails in query_controller
- Stream error chunks when individual target (agent/team/model) execution fails
- Fix TypeScript stream handler to send errors as SSE events instead of JSON responses
- Ensure errors are sent before closing stream so OpenAI SDK and clients receive them

Fixes ARKQB-415: streaming now returns errors instead of silently closing the stream

lucaromagnoli committed:
refactor: address PR comments for streaming error handling

- Replace TODO with GitHub issue #415 reference for streaming improvement
- Extract error types to streaming.py: StreamingErrorDetail and StreamingErrorResponse
- Use typed error structure in proxy_streaming_response
- Remove parseErrorChunk fallback logic, validate StreamError structure directly
- Remove redundant try/catch in ark-cli createChatCompletionStream (errors propagate naturally)

Addresses PR feedback on error handling and type safety

lucaromagnoli committed:
Merge branch 'main' into fix/ARKQB-415-streaming-does-not-return-error

lucaromagnoli committed:
refactor: use expected error structure directly in proxy_streaming_response

- Remove fallback logic for error parsing
- Expect and validate error structure: {"error": {"message": "...", "type": "...", "code": "..."}}
- Only use default error format if structure is completely invalid
- Improves type safety and reduces ambiguity

lucaromagnoli committed:
refactor: use expected error structure directly without fallbacks

- Require both 'message' and 'type' fields in error structure
- Only use default error format if structure is completely invalid
- Remove intermediate variables, build error_data directly from response
- Addresses PR feedback to use expected structure and fail otherwise

lucaromagnoli committed:
fix: reduce nesting complexity in executeTarget error handling

Extract error handling logic into handleTargetExecutionError helper function
to resolve nestif linting error. Also fix gofmt formatting in a2a.go.

havasik commented on ark/internal/controller/query_controller.go:
```diff
@@ -602,6 +614,31 @@ func (r *QueryReconciler) finalize(ctx context.Context, query *arkv1alpha1.Query
 
+// handleTargetExecutionError handles error reporting for target execution failures.
+// It streams errors to clients if streaming is enabled, records telemetry, and emits events.
+func (r *QueryReconciler) handleTargetExecutionError(ctx context.Context, err error, target arkv1alpha1.QueryTarget, span telemetry.Span, metadata map[string]string, eventStream genai.EventStreamInterface, tokenCollector *genai.TokenUsageCollector) {
```
Let's avoid needing to import telemetry - we can handle events recorder and event emission outside of this function consistently

havasik commented on ark/internal/controller/query_controller.go:
```diff
@@ -208,6 +209,17 @@ func (r *QueryReconciler) executeQueryAsync(opCtx context.Context, obj arkv1alph
 	responses, eventStream, err := r.reconcileQueue(opCtx, obj, impersonatedClient, memory, tokenCollector)
 	if err != nil {
+		// Stream error to clients if streaming is enabled
+		if eventStream != nil {
```
Seems like a bit redundant with handleTargetExecutionError

havasik commented on ark/internal/genai/a2a_execution.go:
```diff
@@ -94,6 +94,19 @@ func (e *A2AExecutionEngine) Execute(ctx context.Context, agentName, namespace s
 		})
+
+		if eventStream != nil {
+			modelID := fmt.Sprintf("agent/%s", agentName)
```
same

lucaromagnoli committed:
Merge branch 'main' into fix/ARKQB-415-streaming-does-not-return-error

lucaromagnoli committed:
Merge branch 'main' into fix/ARKQB-415-streaming-does-not-return-error

lucaromagnoli committed:
refactor: address PR comments for streaming error handling

- Remove telemetry dependency from handleTargetExecutionError
- Extract StreamError helper function to avoid code duplication
- Use StreamError helper in query-level, target-level, and A2A error handling
- Move telemetry recording to caller of handleTargetExecutionError
- Remove redundant error streaming code

Addresses @havasik's review comments on PR #372

lucaromagnoli committed:
Merge branch 'main' into fix/ARKQB-415-streaming-does-not-return-error

lucaromagnoli committed:
Merge branch 'main' into fix/ARKQB-415-streaming-does-not-return-error

lucaromagnoli committed:
Merge branch 'main' into fix/ARKQB-415-streaming-does-not-return-error

lucaromagnoli committed:
Merge branch 'main' into fix/ARKQB-415-streaming-does-not-return-error

StarkNDJ commented on services/ark-api/ark-api/src/ark_api/api/v1/openai.py:
```diff
@@ -218,9 +262,9 @@ async def chat_completions(request: ChatCompletionRequest) -> ChatCompletion:
+                    "code": "internal_error",
                 }
-            }
+            },
```
Do we have or want to add style validation across the repo?

lucaromagnoli committed:
Merge branch 'main' into fix/ARKQB-415-streaming-does-not-return-error

lucaromagnoli committed:
Merge branch 'main' into fix/ARKQB-415-streaming-does-not-return-error

dwmkerr commented:
This has been merged in as part of #435

dwmkerr committed:
Merge branch 'main' into fix/ARKQB-415-streaming-does-not-return-error
